### PR TITLE
Sphinx Python Documentation

### PR DESCRIPTION
**Pull request by StefanHabel**
Work in progress.

This PR adds a new build target named `MaterialXDocsPython`, which generates Python API documentation using Sphinx.

The existing developer guide contents are incorporated into the new HTML documentation, which lives side-by-side to the existing Doxygen-generated C++ API documentation.

Docstrings were added to all modules, functions, classes, methods, and attributes.

All parameters were named. None should appear as `arg0`, `arg1`, `arg2` etc anymore.

Statistics about the Python API have been added to the Sphinx build process.

The Sphinx build output currently ends with the following messages, providing an overview of the state of the API docs:
```
The MaterialX Python API consists of:
    * 11 modules
    * 48 functions
    * 139 classes
    * 1180 methods
    * 6 exception types

build succeeded.
```

Docstring example:
```
$ python3 -c 'import PyMaterialXFormat; help(PyMaterialXFormat.flattenFilenames)'
```
```
Help on built-in function flattenFilenames in module PyMaterialXFormat:

flattenFilenames(...) method of builtins.PyCapsule instance
    flattenFilenames(doc: PyMaterialXCore.Document, searchPath: PyMaterialXFormat.FileSearchPath = mx.FileSearchPath(), customResolver: PyMaterialXCore.StringResolver = None) -> None

    Flatten all filenames in the given document, applying string resolvers at
    the scope of each element and removing all fileprefix attributes.

    :param doc: The document to modify.
    :type doc: Document
    :param searchPath: An optional search path for relative to absolute path
        conversion.
    :type searchPath: FileSearchPath
    :param customResolver: An optional custom resolver to apply.
    :type customResolver: StringResolver
```

Preview of generated HTML pages:

Screenshot 2023-10-21 at 14 14 41

<hr>

Screenshot 2023-10-21 at 14 28 03

<hr>

Development environment:

- Apple clang version 15.0.0 (clang-1500.0.40.1)
- Python 3.12.0 (v3.12.0:0fb18b02c8, Oct  2 2023, 09:45:56) [Clang 13.0.0 (clang-1300.0.29.30)] on darwin
- Sphinx 7.2.6
- cmake version 3.27.6

Update #342

### CODE REVIEW

File: documents/sphinx-custom.css [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+    --border-radius-small: 4px;
+    --border-radius-medium: 6px;
+
+    color-scheme: dark;
+
+    --primary-color: #1982d2;
+    --primary-dark-color: #86a9c4;
+    --primary-light-color: #4779ac;
+
+    --box-shadow: 0 2px 8px 0 rgba(0,0,0,.30);
```
Comment (kwokcb) [inline documents/sphinx-custom.css:36]: Doesn't need to be done with this PR, but as "accessibility" is a active area in ASWF, perhaps it's worthwhile to run an accessibility test on the page for the colors and fonts ?

File: documents/CMakeLists.txt [hunk 2]
```diff
@@ -22,13 +23,62 @@ if(DOXYGEN_FOUND)
...
-                       ${CMAKE_SOURCE_DIR}/documents/Images ${CMAKE_CURRENT_BINARY_DIR})
+                       ${CMAKE_SOURCE_DIR}/documents/Images ${DOXYGEN_HTML_OUTPUT_DIR})
     install(DIRECTORY ${DOXYGEN_HTML_OUTPUT_DIR}
             DESTINATION "documents" MESSAGE_NEVER)
+else()
+    message(WARNING "Unable to build MaterialX C++ API Documentation: Doxygen not found.")
 endif(DOXYGEN_FOUND)
+
+
+# MaterialX Python API Documentation
```
Comment (kwokcb) [inline documents/CMakeLists.txt:40]: Wanted to double check if there is any dependence on the actual building the Python modules as that is a build option?

I assume not, but just checking.

File: documents/DeveloperGuide/GraphEditor.md [hunk 3]
```diff
@@ -2,29 +2,30 @@
 
 The MaterialX Graph Editor is an example application for visualizing, creating, and editing MaterialX graphs.  It utilizes the ImGui framework as well as additional ImGui extensions such as the Node Editor.
 
-### Example Images
+## Example Images
+
+![MaterialX Graph Editor with procedural marble example](../Images/MaterialXGraphEditor_Marble.png)
```
Comment (kwokcb) [inline documents/DeveloperGuide/GraphEditor.md:7]: Not specific to this PR but I noticed that images don't show up properly on the `PyPi` package page for MaterialX.
If were adding more images, just curious is there a more robust way to specify image references -- such as an absolute reference to the web site location instead of using relative paths ?

**Review by kwokcb — COMMENTED** [review]
This is a great addition to have !
I left some minor niggly comments. The only real question I had is any inter-dependence on building the actual Python modules.
Thanks.

File: cmake/modules/FindSphinx.cmake [hunk 4]
```diff
@@ -0,0 +1,20 @@
+include(FindPackageHandleStandardArgs)
+
+find_package(Python3)
+if(PYTHON3_FOUND)
```
Comment (jstone-lucasfilm) [inline cmake/modules/FindSphinx.cmake:4]: In case it allows additional simplifications of this logic, I should note that the MaterialX project requires Python 3.6 or greater!

https://github.com/AcademySoftwareFoundation/MaterialX#supported-platforms

File: documents/CMakeLists.txt (hunk 2 above)
Comment (StefanHabel) [reply documents/CMakeLists.txt:55]: Yes indeed, good point -- there is!

I've tweaked this in https://github.com/AcademySoftwareFoundation/MaterialX/pull/1567/commits/9d4b4929c8bcf5f647692f078e44e7956949ab18, to build the Python bindings when Python API documentation is requested to be built:
```cmake
option(MATERIALX_BUILD_PYTHON_DOCS "Create HTML documentation using Sphinx. Requires that Sphinx be installed." OFF)
if (MATERIALX_BUILD_PYTHON_DOCS)
    set(MATERIALX_BUILD_PYTHON ON)
endif()
```
In the same commit, I've also added `DEPENDS` to the custom `MaterialXDocsPython` target:
```cmake
        add_custom_target(
            MaterialXDocsPython
            ${SPHINX_EXECUTABLE}
                -c ${CMAKE_CURRENT_BINARY_DIR}
                ${SPHINX_SOURCE_DIR}
                ${SPHINX_HTML_OUTPUT_DIR}
            WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
            DEPENDS ${MATERIALX_PYTHON_DOCS_TARGET_DEPENDENCIES}
            COMMENT "Building MaterialX Python API Documentation: ${SPHINX_HTML_OUTPUT_DIR}/index.html"
        )
```
The variable that's referenced in `DEPENDS` is a list of targets for all Python C extension modules:
```cmake
        set(
            MATERIALX_PYTHON_DOCS_TARGET_DEPENDENCIES
            MaterialXCore
            PyMaterialXCore
            PyMaterialXFormat
            PyMaterialXGenShader
            PyMaterialXGenGlsl
            PyMaterialXGenOsl
            PyMaterialXGenMdl
            PyMaterialXGenMsl
            PyMaterialXRender
            PyMaterialXRenderGlsl
            PyMaterialXRenderOsl
            PyMaterialXRenderMsl
        )
```

It seems to work well: when making changes to the docstrings in one of the Python C extensions, only that module is rebuilt, and the docs are regenerated.

I'm using the following command line while iterating on the docs, using `rm` to make sure that the docs are rebuilt from scratch (replaced with `RM` below for safety):
```bash
RM -rf  documents/DeveloperGuide/generated/  documents/html-sphinx/  && cmake --build . --target MaterialXDocsPython
```

While it seems to work, I'm happy to explore alternatives, if this isn't quite hitting the mark.

File: documents/DeveloperGuide/index.rst [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+
+Import Order Dependencies
+-------------------------
+
+Note that the order in which the MaterialX Python modules are imported matters, as there are dependencies between them:
+
+- :py:mod:`PyMaterialXCore` -- typically always needed
+    - :py:mod:`PyMaterialXFormat` -- needed for file I/O
+    - :py:mod:`PyMaterialXRender` -- needed for core rendering
```
[the remark this answers is not in the document]
Comment (StefanHabel) [reply documents/DeveloperGuide/index.rst]: Thanks Bernard!

I've actually removed this part from the documentation again, after exploring solutions to the import order dependency issue, and finding something that appears to be working:

- https://github.com/AcademySoftwareFoundation/MaterialX/pull/1567/commits/e1113eabea056260c65ddfa10e66e791f59a840b
- https://github.com/AcademySoftwareFoundation/MaterialX/pull/1567/commits/ab6860bc0a33b42608ccd978d487e3302fb20d09

The approach is to import the required Python modules from the bindings of the Python modules that need them. This is similar to how we'd import required modules in Python code, implemented as a C++ macro definition using `pybind11::module::import()`. The macro supports imports relative to the current module, or as part of the `MaterialX` Python package:
```cpp
// Define a macro to import a PyMaterialX module, e.g. `PyMaterialXCore`,
// either within the `MaterialX` Python package, e.g. in `installed/python/`,
// or as a standalone module, e.g. in `lib/`
#define PYMATERIALX_IMPORT_MODULE(MODULE_NAME)                               \
    try {                                                                    \
        pybind11::module::import("MaterialX." #MODULE_NAME);                 \
    }                                                                        \
    catch (const py::error_already_set&) {                                   \
        pybind11::module::import(#MODULE_NAME);                              \
    }
```
For example, the module code for `PyMaterialXFormat` uses it as follows:
```cpp
    // PyMaterialXFormat depends on types defined in PyMaterialXCore
    PYMATERIALX_IMPORT_MODULE(PyMaterialXCore);
```
While this works in my dev environment (macOS), it is yet to be scrutinized and tested on other platforms.

File: documents/DeveloperGuide/GraphEditor.md (hunk 3 above)
Comment (StefanHabel) [reply documents/DeveloperGuide/GraphEditor.md:7]: Oh, I see the issue: https://pypi.org/project/materialx/

It's great to see that the documentation content is reused in many places!

I've migrated the Markdown docs, including `README.md`, to use absolute paths now:

- https://github.com/AcademySoftwareFoundation/MaterialX/pull/1567/commits/e129d65ee2ba3d3acb9996c131da61c78731d165

File: documents/DeveloperGuide/ShaderGeneration.md [hunk 6]
```diff
@@ -43,12 +43,12 @@ Figure 2. The first option is to keep inline code in a file. The file extension
   <input name="in2" type="color3" />
   <output name="out" type="color3" defaultinput="in1" />
 </nodedef>
-<... more types ...>
+// ... more types ...
```
[the remark this answers is not in the document]
Comment (StefanHabel) [reply documents/DeveloperGuide/ShaderGeneration.md]: Thanks, good point, I've addressed this in https://github.com/AcademySoftwareFoundation/MaterialX/pull/1567/commits/e129d65ee2ba3d3acb9996c131da61c78731d165.

File: documents/sphinx-custom.css (hunk 1 above)
Comment (StefanHabel) [reply documents/sphinx-custom.css:33]: Oh, good point!

Given that we're aiming for some consistency in styles between the various MaterialX sites and docs, could we maybe make this a separate ticket, to look at accessibility across all of the docs?

File: source/MaterialXCore/Material.h [hunk 7]
```diff
@@ -19,7 +19,7 @@ MATERIALX_NAMESPACE_BEGIN
 /// filtered by the given shader type and target.  By default, all surface shader nodes
 /// are returned.
 /// @param materialNode The node to examine.
-/// @param nodeType THe shader node type to return.  Defaults to the surface shader type.
```
Comment (jstone-lucasfilm) [inline source/MaterialXCore/Material.h:22]: @StefanHabel Starting with this edit, I'm seeing a great set of fixes to spelling, letter case, and other details in the C++ libraries for MaterialX.  Would it make sense to separate these out into their own pull request, so that they can be reviewed independently of the new Python Documentation functionality in this proposal?

File: documents/DeveloperGuide/index.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (jstone-lucasfilm) [reply documents/DeveloperGuide/index.rst]: This sounds very promising, @StefanHabel, and this would potentially resolve a dependency issue that we've long wanted to address in MaterialX.

File: documents/CMakeLists.txt (hunk 2 above)
Comment (kwokcb) [reply documents/CMakeLists.txt:55]: This looks good to me. Thanks for looking into it.

File: source/MaterialXCore/Material.h (hunk 7 above)
Comment (StefanHabel) [reply source/MaterialXCore/Material.h:22]: yes, thanks @jstone-lucasfilm !

This PR is getting a little unwieldy.

We could generally split this into separate PRs:

1. Base functionality of generating the Python API Documentation via Sphinx.
2. Fixes in header files for typos encountered while copying docs in headers to docstrings in the Python bindings.
3. Added argument names and docstrings in the Python bindings.

We could treat this PR here (1567) as the one for the Sphinx base functionality, splitting off parts 2 and 3 into separate PRs.

The third part (adding the actual docstrings) could be broken apart further into four separate PRs for the main (groups of) MaterialX Python modules:
- PyMaterialXCore
- PyMaterialXFormat
- PyMaterialXGenShader
    - PyMaterialXGenGlsl
    - PyMaterialXGenOsl
    - PyMaterialXGenMdl
    - PyMaterialXGenMsl
- PyMaterialXRender
    - PyMaterialXRenderGlsl
    - PyMaterialXRenderOsl
    - PyMaterialXRenderMsl

While working on this, I also encountered a couple of issues in the bindings themselves:

- `PyMaterialXCore.UnitDef.setUnitType()` is wrongly bound to `hasUnitType()`: https://github.com/AcademySoftwareFoundation/MaterialX/blob/main/source/PyMaterialX/PyMaterialXCore/PyDefinition.cpp#L76-L77
- `PyMaterialXCore.GeomPropDef`: duplicate definitions can be removed: https://github.com/AcademySoftwareFoundation/MaterialX/blob/main/source/PyMaterialX/PyMaterialXCore/PyGeom.cpp#L60-L71
    - `setGeomProp()`
    - `hasGeomProp()`
    - `getGeomProp()`
- `PyMaterialXGenShader.HwShaderGenerator` static functions are wrongly bound as instance methods: https://github.com/AcademySoftwareFoundation/MaterialX/blob/main/source/PyMaterialX/PyMaterialXGenShader/PyHwShaderGenerator.cpp#L31-L33
    - `bindLightShader()`
    - `unbindLightShader()`
    - `unbindLightShaders()`
- `PyMaterialXRenderMsl.MetalTextureHandler.bindImage()` is wrongly bound to `unbindImage()`: https://github.com/AcademySoftwareFoundation/MaterialX/blob/main/source/PyMaterialX/PyMaterialXRenderMsl/PyMetalTextureHandler.mm#L17-L18

We could treat these issues in the Python bindings as a separate ticket and PR.

Here's a list of tickets and corresponding PRs in which we could organize the overall work here:
- New issue ticket for existing issues in Python bindings
    - New PR to address the issues
- #342
    - #1567 for the Sphinx part
    - Separate new PR for docstrings for PyMaterialXCore
    - Separate new PR for docstrings for PyMaterialXFormat
    - Separate new PR for docstrings for PyMaterialXGen{Shader,Glsl,Osl,Mdl,Msl}
    - Separate new PR for docstrings for PyMaterialXRender{,Glsl,Osl,Msl}
    - Separate new PR for fixes in header files

Alternatively:
- New issue ticket for existing issues in Python bindings
    - New PR to address the issues
- #342
    - #1567 for the Sphinx part and docstrings in all Python modules
    - Separate new PR for fixes in header files

I'd be happy with either approach.

I have more commits to add to this PR here, adding more argument names and docstrings. I'll keeping using this PR here for this for now.

Comment (jstone-lucasfilm) [reply source/MaterialXCore/Material.h:22]: @StefanHabel One approach that could be useful here is to break out one separable change at a time from this larger pull request (e.g. fixes to C++ libraries), so that it can be reviewed and merged to MaterialX main on its own.

As each of these subsets is reviewed and merged to main, we can then click "Update Branch" on this larger pull request, effectively removing that subset from the main proposal.

This should help to focus reviewer attention on a single, independent change at a time, rather than considering the entire proposal at once.

Comment (StefanHabel) [reply source/MaterialXCore/Material.h:22]: Understood! Sounds good!

I'll upload a few more commits to this PR here, to have everything in one place.

I'll then create separate PRs for subsets of this work, which can be reviewed and merged independently.

**Comment by jstone-lucasfilm** [comment]
@StefanHabel We'll be holding ASWF Dev Days 2024 on September 26th and 27th, so let us know if you might be interested in following up on this project then!

https://www.aswf.io/dev-days-2024/

**Comment by StefanHabel** [comment]
> @StefanHabel We'll be holding ASWF Dev Days 2024 on September 26th and 27th, so let us know if you might be interested in following up on this project then!
> 
> https://www.aswf.io/dev-days-2024/

Absolutely, count me in!

Apologies for putting this on hold for so long.

Looking forward to it!

File: documents/DeveloperGuide/GraphEditor.md (hunk 3 above)
Comment (StefanHabel) [reply documents/DeveloperGuide/GraphEditor.md:7]: The update of these markdown files has been split off into its own PR: https://github.com/AcademySoftwareFoundation/MaterialX/pull/2037

**Comment by jstone-lucasfilm** [comment]
Thanks for getting this started, @StefanHabel, and I believe we can close out this earlier PR in favor of your recent work in #2038.
